milatools/utils/compute_node.py

```diff
@@ -0,0 +1,512 @@
+    """Get the IDS of jobs created by milatools from the output of `sacct`."""
+    jobs = await login_node.get_output_async(
+        f"sacct --noheader --allocations --user=$USER "
+        f"--starttime=now-{int(since.total_seconds())}seconds --format=JobId "
+        f"--name={job_name}"
+    )
+    return set([int(job_id_str) for job_id_str in jobs.splitlines()])
+
+
+@contextlib.asynccontextmanager
+async def cancel_new_jobs_on_interrupt(login_node: RemoteV2, job_name: str):
+    """ContextManager that handles interruptions while creating a new allocation.
+
+    This handles the case where an interrupt is raised while running a command over SSH
+    that creates a new job allocation (either salloc or sbatch) before we are able to
+    parse the job id. (In the case where we have the job ID, we simply use
+    `scancel {job_id}`).
+
+    In this case, we try to cancel the new job(s) that have appeared since entering the
+    `async with` block that have the name `job_name`. Emits a warning in the (unlikely)
+    case where such jobs are not found, as it means that there could be a "zombie" job
+    allocation on the cluster.
+    """
+    jobs_before = await get_queued_milatools_job_ids(login_node, job_name=job_name)
+    if jobs_before:
+        logger.info(
+            f"Existing jobs on {login_node.hostname} with name {job_name}: {jobs_before}"
+        )
+    else:
+        logger.debug(
+            f"There are currently no jobs with name {job_name} on {login_node.hostname}."
+        )
+    try:
+        yield
+    except (KeyboardInterrupt, asyncio.CancelledError):
+        jobs_after = await get_queued_milatools_job_ids(login_node, job_name=job_name)
+        logger.warning("Interrupted before we were able to parse a job id!")
+        # We were unable to get the job id, so we'll try to cancel only the newly
+        # spawned jobs from this user that match the set name.
+        new_jobs = list(set(jobs_after) - set(jobs_before))
+        if len(new_jobs) == 1:
+            job_id = new_jobs[0]
+            console.log(
+                f"Cancelling job {job_id} since it is the only new job of this "
+                f"user with name {job_name!r} since the last call to salloc or sbatch.",
+                style="yellow",
+            )
+            login_node.run(f"scancel {new_jobs[0]}", display=True, hide=False)
+        elif len(new_jobs) > 1:
+            console.log(
+                f"There appears to be more than one new jobs from this user with "
+                f"name {job_name!r} since the initial call to salloc or sbatch: "
+                f"{new_jobs}\n"
+                "Cancelling all of them to be safe...",
+                style="yellow",
+            )
+            await asyncio.shield(
+                login_node.run_async(
+                    "scancel " + " ".join(str(job_id) for job_id in new_jobs),
+                    display=True,
+                    hide=False,
+                )
+            )
```

Comment:
(Assuming that this comment is with respect to `cancel_new_jobs_on_interrupt`, since things appear to have moved a bit). 
Indeed, you're right, asyncio.shield + await run_async makes no sense. Might as well just do `run` (synchronously) in that case.
Fixed in 837541161a780bd3af0f4e3e612f8232366c15ea.